### PR TITLE
Don't use external service tokens in batch changes

### PR DESCRIPTION
**Pull request by eseliger**
Closes https://github.com/sourcegraph/sourcegraph/issues/16814

### CODE REVIEW

File: enterprise/internal/batches/sources/sources.go [hunk 1]
```diff
@@ -245,6 +226,7 @@ func WithSiteAuthenticator(ctx context.Context, tx SourcerStore, css ChangesetSo
 	if cred != nil {
 		return css.WithAuthenticator(cred)
 	}
+	// TODO: This should return ErrMissingCredentials.
```
Comment (eseliger) [inline enterprise/internal/batches/sources/sources.go:229]: I'll do that in the future when we're ready to not use external service tokens for the changeset syncer.

File: enterprise/internal/batches/sources/sources.go [hunk 2]
```diff
@@ -114,19 +114,23 @@ func (s *sourcer) loadBatchesSource(ctx context.Context, tx SourcerStore, extern
...
+	// 	return css.WithAuthenticator(cred)
 	// }
 	return css, nil
 }
 
 func gitserverPushConfig(ctx context.Context, store *database.ExternalServiceStore, repo *types.Repo, au auth.Authenticator) (*protocol.PushConfig, error) {
+	// Empty authenticators are not allowed.
+	if au == nil {
+		return nil, ErrNoPushCredentials{}
+	}
```
Comment (eseliger) [inline enterprise/internal/batches/sources/sources.go:132]: This is the change for pushing a branch to the code host

File: enterprise/internal/batches/sources/sources.go [hunk 3]
```diff
@@ -213,22 +210,6 @@ func WithAuthenticatorForUser(ctx context.Context, tx SourcerStore, css Changese
...
-	// not, then we need to error out.
-	// Once we tackle https://github.com/sourcegraph/sourcegraph/issues/16814,
-	// this code path should be removed.
-	user, err := database.Users(tx.DB()).GetByID(ctx, userID)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to load user")
-	}
-	if user.SiteAdmin {
-		return css, nil
-	}
```
Comment (eseliger) [inline enterprise/internal/batches/sources/sources.go:231]: ..and this is the change for interacting with the APIs of the code hosts

File: enterprise/internal/batches/testing/rsa.go [hunk 4]
```diff
@@ -18,3 +22,24 @@ func MockRSAKeygen(t *testing.T) {
 		encryption.MockGenerateRSAKey = nil
 	})
 }
+
+type createSiteCredentialer interface {
+	CreateSiteCredential(context.Context, *btypes.SiteCredential, auth.Authenticator) error
+}
+
+func CreateTestSiteCredential(t *testing.T, bstore createSiteCredentialer, repo *types.Repo) *btypes.SiteCredential {
```
Comment (LawnGnome) [inline enterprise/internal/batches/testing/rsa.go:30]: Probalby not worth changing, but this doesn't feel like it belongs in a file called `rsa.go` to me.

**Pull request merged**